File: relationship-manager/wealth-projection-service/src/main/java/com/bank/rm/projection/service/MonteCarloEngine.java
package com.bank.rm.projection.service;

import com.bank.rm.common.dto.RiskCategory;
import com.bank.rm.projection.dto.ProjectionDtos.*;
import org.apache.commons.math3.distribution.NormalDistribution;
import java.util.*;
import java.util.concurrent.*;
import java.util.stream.IntStream;

public class MonteCarloEngine {
    private static final int TOTAL_SCENARIOS = 10_000;
    private static final int BATCH_COUNT = 4;
    private static final double INFLATION_RATE = 0.06;

    private static final Map<String, double[]> ASSET_PARAMS = Map.of(
        "equity", new double[]{0.12, 0.18},
        "debt", new double[]{0.07, 0.04},
        "gold", new double[]{0.08, 0.12},
        "cash", new double[]{0.04, 0.01},
        "alternative", new double[]{0.15, 0.25}
    );

    public ProjectionResult runProjection(ProjectionInput input) {
        Map<String, Double> allocation = getAllocation(input.riskCategory());
        ExecutorService executor = Executors.newFixedThreadPool(BATCH_COUNT);
        int perBatch = TOTAL_SCENARIOS / BATCH_COUNT;
        List<Future<double[]>> futures = new ArrayList<>();

        for (int b = 0; b < BATCH_COUNT; b++) {
            int seed = b;
            futures.add(executor.submit(() -> runBatch(perBatch, input, allocation, seed)));
        }

        double[][] allFinalValues = new double[TOTAL_SCENARIOS][];
        double[][] yearlyMedians = new double[input.years()][TOTAL_SCENARIOS];
        int idx = 0;
        try {
            for (Future<double[]> f : futures) {
                double[] batchResults = f.get();
                for (int i = 0; i < perBatch; i++) {
                    allFinalValues[idx] = new double[]{batchResults[i]};
                    idx++;
                }
            }
        } catch (Exception e) {
            throw new RuntimeException("Monte Carlo computation failed", e);
        } finally {
            executor.shutdown();
        }

        double[] finalValues = Arrays.stream(allFinalValues).mapToDouble(a -> a[0]).toArray();
        Arrays.sort(finalValues);

        PercentileValues nominal = new PercentileValues(
            percentile(finalValues, 10), percentile(finalValues, 25),
            percentile(finalValues, 50), percentile(finalValues, 75),
            percentile(finalValues, 90)
        );

        double inflationFactor = Math.pow(1 + INFLATION_RATE, input.years());
        PercentileValues real = new PercentileValues(
            nominal.p10() / inflationFactor, nominal.p25() / inflationFactor,
            nominal.p50() / inflationFactor, nominal.p75() / inflationFactor,
            nominal.p90() / inflationFactor
        );

        double targetProb = input.targetAmount() > 0 ?
            (Arrays.stream(finalValues).filter(v -> v >= input.targetAmount()).count() * 100.0 / TOTAL_SCENARIOS) : 0;

        return new ProjectionResult(nominal, real, TOTAL_SCENARIOS, Math.round(targetProb * 10.0) / 10.0);
    }

    private double[] runBatch(int count, ProjectionInput input, Map<String, Double> allocation, int seed) {
        Random random = new Random(seed * 42L);
        double[] results = new double[count];
        for (int s = 0; s < count; s++) {
            double portfolio = input.initialInvestment();
            for (int y = 0; y < input.years(); y++) {
                double yearReturn = 0;
                for (Map.Entry<String, Double> entry : allocation.entrySet()) {
                    double[] params = ASSET_PARAMS.getOrDefault(entry.getKey(), new double[]{0.05, 0.10});
                    NormalDistribution dist = new NormalDistribution(params[0], params[1]);
                    double assetReturn = dist.inverseCumulativeProbability(random.nextDouble());
                    yearReturn += entry.getValue() * assetReturn;
                }
                portfolio = portfolio * (1 + yearReturn) + input.annualContribution();
            }
            results[s] = Math.max(0, portfolio);
        }
        return results;
    }

    private Map<String, Double> getAllocation(RiskCategory category) {
        return switch (category) {
            case CONSERVATIVE -> Map.of("equity", 0.20, "debt", 0.50, "gold", 0.15, "cash", 0.10, "alternative", 0.05);
            case MODERATE -> Map.of("equity", 0.40, "debt", 0.30, "gold", 0.15, "cash", 0.05, "alternative", 0.10);
            case AGGRESSIVE -> Map.of("equity", 0.60, "debt", 0.15, "gold", 0.10, "cash", 0.05, "alternative", 0.10);
            case VERY_AGGRESSIVE -> Map.of("equity", 0.75, "debt", 0.05, "gold", 0.05, "cash", 0.00, "alternative", 0.15);
        };
    }

    private double percentile(double[] sorted, int p) {
        int index = (int) Math.ceil(p / 100.0 * sorted.length) - 1;
        return Math.round(sorted[Math.max(0, index)] * 100.0) / 100.0;
    }
}

File: relationship-manager/wealth-projection-service/src/main/kotlin/com/bank/rm/projection/service/MonteCarloEngine.kt
package com.bank.rm.projection.service

import com.bank.rm.common.dto.RiskCategory
import org.apache.commons.math3.distribution.NormalDistribution
import org.slf4j.LoggerFactory
import org.springframework.stereotype.Component
import java.util.concurrent.CompletableFuture
import java.util.concurrent.Executors
import kotlin.math.pow
import kotlin.math.sqrt

/**
 * Monte Carlo simulation engine for wealth projection.
 * Runs 10,000 scenarios to model portfolio growth with stochastic returns.
 *
 * Per LLD Section 7: Uses asset-class return assumptions, volatility,
 * and correlation to simulate realistic portfolio outcomes.
 */
@Component
class MonteCarloEngine {

    private val log = LoggerFactory.getLogger(javaClass)
    private val executor = Executors.newFixedThreadPool(4)

    companion object {
        const val DEFAULT_SCENARIOS = 10_000
        const val INFLATION_RATE = 0.06 // 6% India CPI
    }

    // Asset class return assumptions (annualized)
    private val assetReturns = mapOf(
        "equity" to AssetAssumption(expectedReturn = 0.12, volatility = 0.18),
        "debt" to AssetAssumption(expectedReturn = 0.07, volatility = 0.04),
        "gold" to AssetAssumption(expectedReturn = 0.08, volatility = 0.12),
        "cash" to AssetAssumption(expectedReturn = 0.04, volatility = 0.01),
        "alternative" to AssetAssumption(expectedReturn = 0.15, volatility = 0.25)
    )

    fun runProjection(input: ProjectionInput): ProjectionResult {
        log.info("Running Monte Carlo simulation: {} scenarios, {} years",
            input.scenarios, input.projectionYears)

        val allocation = input.portfolioAllocation
        val scenarios = input.scenarios

        // Run simulation in parallel batches
        val batchSize = scenarios / 4
        val futures = (0 until 4).map { batch ->
            CompletableFuture.supplyAsync({
                simulateBatch(input, batch * batchSize, batchSize)
            }, executor)
        }

        val allFinalValues = futures.flatMap { it.get() }
        val sorted = allFinalValues.sorted()

        // Calculate percentiles
        val p10 = percentile(sorted, 10)
        val p25 = percentile(sorted, 25)
        val p50 = percentile(sorted, 50)
        val p75 = percentile(sorted, 75)
        val p90 = percentile(sorted, 90)

        // Calculate year-by-year median trajectory
        val yearlyMedian = calculateYearlyMedian(input)

        // Calculate probability of reaching target
        val targetProbability = if (input.targetAmount > 0) {
            allFinalValues.count { it >= input.targetAmount }.toDouble() / scenarios * 100
        } else 0.0

        // Real (inflation-adjusted) values
        val inflationFactor = (1 + INFLATION_RATE).pow(input.projectionYears)
        val realP50 = p50 / inflationFactor

        return ProjectionResult(
            nominalPercentiles = PercentileValues(p10, p25, p50, p75, p90),
            realPercentiles = PercentileValues(
                p10 / inflationFactor, p25 / inflationFactor,
                realP50, p75 / inflationFactor, p90 / inflationFactor
            ),
            yearlyProjection = yearlyMedian,
            targetAmount = input.targetAmount,
            probabilityOfTarget = Math.round(targetProbability * 10.0) / 10.0,
            meanFinalValue = allFinalValues.average(),
            scenarios = scenarios,
            projectionYears = input.projectionYears,
            inflationRate = INFLATION_RATE
        )
    }

    private fun simulateBatch(input: ProjectionInput, startIdx: Int, count: Int): List<Double> {
        val results = mutableListOf<Double>()
        val normal = NormalDistribution()

        for (i in 0 until count) {
            var portfolio = input.initialInvestment

            for (year in 1..input.projectionYears) {
                // Generate correlated returns for each asset class
                val yearReturn = calculatePortfolioReturn(input.portfolioAllocation, normal)
                portfolio = portfolio * (1 + yearReturn) + input.annualContribution
            }

            results.add(portfolio)
        }

        return results
    }

    private fun calculatePortfolioReturn(
        allocation: Map<String, Double>,
        normal: NormalDistribution
    ): Double {
        var portfolioReturn = 0.0

        for ((assetClass, weight) in allocation) {
            val assumption = assetReturns[assetClass] ?: continue
            val randomReturn = assumption.expectedReturn + assumption.volatility * normal.sample()
            portfolioReturn += weight / 100.0 * randomReturn
        }

        return portfolioReturn
    }

    private fun calculateYearlyMedian(input: ProjectionInput): List<YearlyProjection> {
        val normal = NormalDistribution()
        val miniScenarios = 1000
        val results = Array(input.projectionYears) { mutableListOf<Double>() }

        repeat(miniScenarios) {
            var portfolio = input.initialInvestment
            for (year in 0 until input.projectionYears) {
                val yearReturn = calculatePortfolioReturn(input.portfolioAllocation, normal)
                portfolio = portfolio * (1 + yearReturn) + input.annualContribution
                results[year].add(portfolio)
            }
        }

        return results.mapIndexed { idx, values ->
            val sorted = values.sorted()
            val inflationFactor = (1 + INFLATION_RATE).pow(idx + 1)
            YearlyProjection(
                year = idx + 1,
                nominalMedian = percentile(sorted, 50),
                realMedian = percentile(sorted, 50) / inflationFactor,
                p25 = percentile(sorted, 25),
                p75 = percentile(sorted, 75)
            )
        }
    }

    private fun percentile(sorted: List<Double>, percentile: Int): Double {
        val index = (percentile / 100.0 * (sorted.size - 1)).toInt()
        return sorted[index]
    }
}

data class AssetAssumption(
    val expectedReturn: Double,
    val volatility: Double
)

data class ProjectionInput(
    val initialInvestment: Double,
    val annualContribution: Double,
    val portfolioAllocation: Map<String, Double>,
    val projectionYears: Int,
    val targetAmount: Double = 0.0,
    val scenarios: Int = MonteCarloEngine.DEFAULT_SCENARIOS
)

data class ProjectionResult(
    val nominalPercentiles: PercentileValues,
    val realPercentiles: PercentileValues,
    val yearlyProjection: List<YearlyProjection>,
    val targetAmount: Double,
    val probabilityOfTarget: Double,
    val meanFinalValue: Double,
    val scenarios: Int,
    val projectionYears: Int,
    val inflationRate: Double
)

data class PercentileValues(
    val p10: Double,
    val p25: Double,
    val p50: Double,
    val p75: Double,
    val p90: Double
)

data class YearlyProjection(
    val year: Int,
    val nominalMedian: Double,
    val realMedian: Double,
    val p25: Double,
    val p75: Double
)
